setup.cfg

```diff
@@ -42,15 +42,15 @@ where = src
 
 [options.extras_require]
 dev =
-    pre-commit
-    tox
+    pre-commit == 2.21.0
+    tox == 4.0.1
 docs =
-    pyenchant
-    sphinx
-    sphinx-autobuild
-    sphinx-rtd-theme
-    sphinx-tabs
-    sphinxcontrib-spelling
+    furo == 2022.12.7
+    pyenchant == 3.2.2
+    sphinx == 6.1.2
+    sphinx_tabs == 3.4.1
+    sphinx-autobuild == 2021.3.14
+    sphinxcontrib-spelling == 7.7.0
```

Comment:
Since it's not really documented elsewhere, here's the reasoning. The intention behind this set of changes across projects is to try and provide:
* For end users, an runtime environment that is accomodating as possible, while providing pre-emptive protection against backwards incompatible changes. At least in theory, semantic versioning should only increase a major if there's a backwards incompatible API change, the pinning strategy here means that we have to opt into that change.
* For developers/CI, an environment that is a specific as possible, without requiring multiple specification of requirements (which is effectively what lock files or external requirements files represent).

Regarding the specific points you've made:

1. The pinning policy is "oldest version that meets API requirements, but prevent major bumps". The upper limit is only imposed for packages that (a) apparently use semver and (b) aren't part of the "core" Python ecosystem (tools like pip, wheel, etc). The "oldest" versions haven't been exhaustively tested; they're "old enough for practical purposes". If someone wanted to propose a fix proposing an earlier version for some reason, the theory is we'd accept that change as long as it passed CI.
2. Yes, this pins rubicon-objc to tox 4.2.8 precisely. However, this should only be an issue if you're sharing your rubicon-objc development environment with another project, which probably isn't a good idea anyway. If you've got a specific rubicon-objc development environment, you shouldn't have any clashes; and in the Toga environment, you'll have the *Toga* development dependencies, not the Rubicon ones.
3. You're correct that it doesn't pin *all* dependencies; however, there's a "practicality vs purity" balance here. What we're doing here is specifying the public APIs that *we* are dependent on, and relying on end projects to specify their own runtime dependencies. If it ever turns out that we have a transitive dependency on a specific version of docutils, I'd argue either (a) that's a bug in sphinx, because it hasn't correctly captured it's dependencies; or (b) we *are* using a feature of docutils, in which case we *do* have a dependency that should be declared; or (c) there's a bug in docutils, in which case we can add a specific requirement as a workaround for the issue (and document that it is a workaround)
4. There *is* a way to auto-update, and it's already in place: dependabot. Every Sunday, all the Python dependencies are checked, and if new versions exist, they're bumped - e.g., [this update](https://github.com/beeware/toga/pull/1739/files) from last week on Toga. Dependabot checks all the install_requires and extra_requires; if there's a new version, it either updates the specifically pinned version, or if there's a new major version, proposes a new major API bump. Each of these changes is done in a standalone PR, so we get an independent CI run for each specific change, which gives us a chance to see which specific version bump caused a problem.

Adding lockfiles or requirements files would be another approach that could work; however, it would come at a cost:
1. dependabot would become exceedingly difficult to work with - the long tail of dependencies would mean we'd have a PR for every micro version in a dependency 4 transient relationships deep.
2. We would be, at some level, duplicating details. We'd have an extras definition providing one set of version specifiers, and requirements/lock file specifying another. I guess the two *should* remain compatible under the scheme you've proposed; but the development advice (and CI practice) would be to always use the requirements/lock file - so we wouldn't actually benefit from having the specifier in extras_requires.

For me, this comes down to practicality over purity. Yes, a requirements/lockfile would give us a hard definition of a specific environment - but I don't think that's actually to our advantage here. If were deploying something to production, I'd 100% agree - that should be pinned to every version that could be involved. But we're talking about a development environment, where there is benefit in known stability, but only stability in what we are *explicitly* dependant upon. Beyond that, we'd be doing accounting for specifics that we don't *really* care about.

Reply:
I agree: the main requirements list needs to be able to coexist with any other reasonably-recent package, but the dev requirements will only be used by Rubicon's own developers, so it's OK if they're more specific, and this will help make the builds and tests more reproducible.

Reply:
Thanks for the detailed explanation. I missed that we already have Dependabot configured for Python dependencies as well - that solves the updating problem obviously.

Yeah, since only the dev/docs dependencies are pinned to exact versions, it won't affect reguar users, and for development you'll *usually* use separate virtual environments. It only really matters when you're debugging cross-library issues and have to install multiple libraries in "dev mode" into the same environment, but that's relatively rare.

I agree on the "practicality beats purity" part, but I suppose I have different opinions on what's practical :) When not going for a 100% reproducible build with a full lock file, I feel like it's more practical to save yourself the effort of micro-managing versions and just let minor updates happen freely. After all, you're already trusting the *indirect* deps to not break their API within some version range, so why not trust the direct deps the same way?

But as I said, I don't mean to push back hard here. It's always a tradeoff, and this solution does indeed make things *more* reproducible.

Reply:
An observation regarding the "multiple libraries in dev mode" case: dependabot updates everything at the same cadence. This morning, we had 2 pull requests opened: [toga#1748](https://github.com/beeware/toga/pull/1748) and [briefcase#1057](https://github.com/beeware/briefcase/pull/1057) - both updating to a new version of Tox. So, as long as you branch projects at the same point in time, and the projects are all using the same "pinned dev dependencies" strategies, you *should* have compatible development environments as well.